<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления" data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления</a></span></li><li><span><a href="#Изменение-аномалий-и-типов-данных" data-toc-modified-id="Изменение-аномалий-и-типов-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Изменение аномалий и типов данных</a></span></li><li><span><a href="#Удаление-дубликатов" data-toc-modified-id="Удаление-дубликатов-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Удаление дубликатов</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма" data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма</a></span></li><li><span><a href="#Категоризация-дохода" data-toc-modified-id="Категоризация-дохода-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Категоризация дохода</a></span></li><li><span><a href="#Категоризация-целей-кредита" data-toc-modified-id="Категоризация-целей-кредита-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Категоризация целей кредита</a></span></li><li><span><a href="#Формирование-сводных-таблиц" data-toc-modified-id="Формирование-сводных-таблиц-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Формирование сводных таблиц</a></span><ul class="toc-item"><li><span><a href="#Зависимость-между-количеством-детей-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-количеством-детей-и-возвратом-кредита-в-срок-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Зависимость между количеством детей и возвратом кредита в срок</a></span></li><li><span><a href="#Зависимость-между-семейным-положением-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-семейным-положением-и-возвратом-кредита-в-срок-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>Зависимость между семейным положением и возвратом кредита в срок</a></span></li><li><span><a href="#Зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span>Зависимость между уровнем дохода и возвратом кредита в срок</a></span></li><li><span><a href="#Влияние-целей-кредита-на-его-возврат-в-срок" data-toc-modified-id="Влияние-целей-кредита-на-его-возврат-в-срок-9.4"><span class="toc-item-num">9.4&nbsp;&nbsp;</span>Влияние целей кредита на его возврат в срок</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Исследование надежности заёмщиков

<span style="color:green"><font size="3"><b>1. Необходимо проверить влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.</b></span>

<span style="color:orange"><font size="3"><b>2. Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.</b></font></span>

<span style="color:brown"><font size="3"><b>3. Входные данные от банка — статистика о платёжеспособности клиентов.</b></span>

## Обзор данных

In [65]:
# 1. Подключение модуля pathlib
import pathlib
# 2. Подключение класса Path из модуля pathlib
from pathlib import Path

# 3. Подключение библиотеки Pandas
import pandas as pd

In [66]:
# Формирование и вывод на экран датафрейма df
# с отработкой ошибок try-except
try:
    df = pd.read_csv('/datasets/data_credit.csv')
    display(df)
except:
    df = pd.read_csv(
        Path(
            pathlib.Path.cwd(),
            'datasets',
            'data_credit.csv'
        )
    )
    display(df)

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

In [67]:
# Вывод списка характерных значений в столбцах датафрейма df
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

In [68]:
# Получение информации по названиям столбцов df
df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

In [69]:
# Получение общей информации по таблице из df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [70]:
# Определение количества пропусков
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 1:</b></span>
    
<span style="color:steelblue"><font size="2"><b>1. Названия колонок таблицы написаны с соблюдением змеиного стиля и не имеют пропусков.\
    2. Таблица состоит из 12 колонок и 21525 строк.\
    3.1. В колонках 'days_employed' и 'total_income' есть 2174 пропуска (ориентировочно 10% от общего количества, так как количество не нулевых значений равно 19351).\
    Пропуски необходимо заменить медианными значениями в виду завышения дохода в 'total_income' при использовании среднего арифметического значения.\
    3.2. Значения в колонках 'days_employed' и 'total_income' представлены в вещественном типе float64.\
    Необходимо заменить на числовой тип int64.</b></span></div>

## Заполнение пропусков

In [71]:
# Заполнение пропусков в колонках 'days_employed' и 'total_income' медианным
# значением в соответствии с источником дохода в 'income_type'.
# Создание функции для заполнения пропусков медианным значением:
# 1. Объявление функции without_nan()
def without_nan(columns):
    # 2. Вложенный цикл для заполнения пропусков NaN в колонках
    # и для перебора названий колонок из списка columns_list
    for value in columns:
        # 2.1. Заполнение пропусков NaN нулями
        df[value] = df[value].fillna(0)
        # 2.2. Создание списка уникальных видов дохода
        income_type_list = df['income_type'].unique()
        # 3. Вложенный цикл второго периода для перебора
        # видов дохода из списка income_type_list
        # и формирования датафреймов при соблюдении
        # условий логической индексации с дальнейшим
        # вычислением медианных значений в необходимых
        # столбцах value полученных датафреймов
        for string in income_type_list:
            # 3.1. Формирование датафреймов с помощью
            # логической индексации по значению string
            df_string = df[df['income_type'] == string]
            # 3.2. Получение медианного значения столбца
            # value в сформированном датафрейме
            df_median = df_string[value].median()
            # 3.3. Замена нулевых значений в колонках
            # value полученными медианными значениями
            df.loc[(df[value] == 0) & (df['income_type'] == string), value] = df_median

In [72]:
# Список колонок с пропусками
columns_list = ['days_employed', 'total_income']

In [73]:
# Применение функции without_nan() к списку columns_list для заполнения пропусков
# с обработкой ошибок try-except
try:
    without_nan(columns_list)
except:
    print('Проверь алгоритм работы функции "without_nan()"!')

## Проверка данных на аномалии и исправления

In [74]:
# Проверка отсутствия пропусков после вызова функции without_nan()
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [75]:
# Определение количества явных дубликатов
df.duplicated().sum()

54

In [76]:
# Просмотр датафрейма с явными дубликатами
df[df.duplicated()]

children  days_employed  dob_years education  education_id  \
2849          0   -1360.363902         41   среднее             1   
4182          1   -1360.363902         34    ВЫСШЕЕ             0   
4851          0  360505.668544         60   среднее             1   
5557          0  360505.668544         58   среднее             1   
7808          0  360505.668544         57   среднее             1   
8583          0  360505.668544         58    высшее             0   
9238          2   -1360.363902         34   среднее             1   
9528          0  360505.668544         66   среднее             1   
9627          0  360505.668544         56   среднее             1   
10462         0  360505.668544         62   среднее             1   
10697         0   -1311.128244         40   среднее             1   
10864         0  360505.668544         62   среднее             1   
10994         0  360505.668544         62   среднее             1   
11791         0   -1360.363902         47   среднее             1   
12373         0   -1360.363902         58   среднее             1   
12375         1   -1360.363902         37   среднее             1   
12736         0  360505.668544         59   среднее             1   
13025         1   -1360.363902         44   среднее             1   
13639         0  360505.668544         64   среднее             1   
13773         0   -1360.363902         35   среднее             1   
13878         1   -1311.128244         31   среднее             1   
13942         0   -1360.363902         44   среднее             1   
14432         2   -2385.358043         36    высшее             0   
14832         0   -1360.363902         50   среднее             1   
15091         0   -1360.363902         58   среднее             1   
15188         0   -1360.363902         60   среднее             1   
15273         0  360505.668544         57   среднее             1   
16176         0   -1360.363902         47   среднее             1   
16378         0   -1360.363902         46   среднее             1   
16902         2   -1360.363902         39   среднее             1   
16904         1   -1360.363902         32    высшее             0   
17379         0   -1311.128244         54    высшее             0   
17755         1   -1360.363902         43   среднее             1   
17774         1   -1311.128244         40   среднее             1   
18328         0   -1360.363902         29    высшее             0   
18349         1   -2385.358043         30    высшее             0   
18428         0  360505.668544         64   среднее             1   
18521         0   -1360.363902         56   среднее             1   
18563         0   -1360.363902         54   среднее             1   
18755         0  360505.668544         58   среднее             1   
19041         0   -1360.363902         56   среднее             1   
19184         0   -1360.363902         46   среднее             1   
19321         0   -1360.363902         23   среднее             1   
19387         0   -1311.128244         38    высшее             0   
19688         0  360505.668544         61   среднее             1   
19832         0   -1360.363902         48   среднее             1   
19946         0   -1360.363902         57   среднее             1   
20116         0  360505.668544         57   среднее             1   
20165         0   -1360.363902         42   среднее             1   
20702         0  360505.668544         64   среднее             1   
21032         0  360505.668544         60   среднее             1   
21132         0   -1360.363902         47   среднее             1   
21281         1   -1360.363902         30    высшее             0   
21415         0  360505.668544         54   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
2849         женат / замужем                 0      F    сотрудник     0   
4182        гражданский брак                 1      F    со

In [77]:
# Поиск неявных дубликатов в 'children'
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [78]:
# Вывод списка характерных значений в столбце 'days_employed'
df['days_employed'].describe()

count     21525.000000
mean      63478.813196
std      140950.114760
min      -18388.949901
25%       -2518.168900
50%       -1311.128244
75%        -316.061820
max      401755.400475
Name: days_employed, dtype: float64

In [79]:
# Вывод списка характерных значений в столбце 'dob_years'
df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [80]:
# Просмотр датафрейма с нулевым значением в 'dob_years'
df[df['dob_years'] == 0]

children  days_employed  dob_years education  education_id  \
99            0  346541.618895          0   Среднее             1   
149           0   -2664.273168          0   среднее             1   
270           3   -1872.663186          0   среднее             1   
578           0  397856.565013          0   среднее             1   
1040          0   -1158.029561          0    высшее             0   
...         ...            ...        ...       ...           ...   
19829         0   -1360.363902          0   среднее             1   
20462         0  338734.868540          0   среднее             1   
20577         0  331741.271455          0   среднее             1   
21179         2    -108.967042          0    высшее             0   
21313         0   -1268.487728          0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
99           женат / замужем                 0      F   пенсионер     0   
149                в разводе                 3      F   сотрудник     0   
270          женат / замужем                 0      F   сотрудник     0   
578          женат / замужем                 0      F   пенсионер     0   
1040               в разводе                 3      F   компаньон     0   
...                      ...               ...    ...         ...   ...   
19829        женат / замужем                 0      F   сотрудник     0   
20462        женат / замужем                 0      F   пенсионер     0   
20577  Не женат / не замужем                 4      F   пенсионер     0   
21179        женат / замужем                 0      M   компаньон     0   
21313  Не женат / не замужем                 4      M   сотрудник     0   

        total_income                                 purpose  
99      71291.522491                              автомобиль  
149     70176.435951                       операции с жильем  
270    102166.458894                            ремонт жилью  
578     97620.687042  строительство собственной недвижимости  
1040   303994.134987                         свой автомобиль  
...              ...                                     ...  
19829  133546.457238                                   жилье  
20462  259193.920299                    покупка своего жилья  
20577  129788.762899                            недвижимость  
21179  240702.007382        строительство жилой недвижимости  
21313  152419.184851                       операции с жильем  

[101 rows x 12 columns]

In [81]:
# Поиск неявных дубликатов в 'education'
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

In [82]:
# Поиск неявных дубликатов в 'family_status'
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [83]:
# Поиск неявных дубликатов в 'gender'
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [84]:
# Вывод строки со значением 'XNA' в 'gender'
df[df['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0   -2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

In [85]:
# Поиск неявных дубликатов в 'debt'
df['debt'].unique()

array([0, 1], dtype=int64)

In [86]:
# Вывод списка характерных значений в столбце 'total_income'
df['total_income'].describe()

count    2.152500e+04
mean     1.642764e+05
std      9.823691e+04
min      2.066726e+04
25%      1.077982e+05
50%      1.398982e+05
75%      1.955499e+05
max      2.265604e+06
Name: total_income, dtype: float64

In [87]:
# Поиск неявных дубликатов в 'purpose'
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 2:</b></span>

<span style="color:steelblue"><font size="2"><b>1. Функция 'without_nan()' заполнила пропуски медианными значениями.\
    2. Выявлено 54 явных дубликата методом '.duplicated()'.\
    3.1. В колонке 'children' обнаружено отрицательное значение -1.\
    Количество детей не может быть меньше 0. Возможно при внесении анкетных данных была допущена ошибка в виде знака '-' перед цифрой 1. Так как количество таких значений 47 и составляет малую часть, то в рамках проекта -1 заменяю на 1.\
    3.2. Также вызывает сомнение количество детей 20. Возможно случайно набрали либо 0, либо 2. Так как цифра положительная, а в России/мире существуют многодетные семьи, то в рамках проекта оставляю данное значение.\
    3.3. Данные необходимо уточнить у заказчика.\
    4.1. В колонке 'days_employed' числа представлены с отрицательным значением. Исключение составляют значения для пенсионеров - числа положительные и аномально большие.\
    Медиана для колонки с положительными и отрицательными значениями будет некорректной.\
    4.2. Появление отрицательных значений возможно связано с методикой расчёта стажа - в системе unix time от колонки с меньшим числом (дата/возраст начала трудового стажа) отнимают колонку с большим числом (дата/возраст на момент подачи заявления на получение кредита).\
    4.3. Появление положительных и аномально больших значений трудового стажа для пенсионеров возможно связано с иным методом расчёта в банковской системе.\
    4.4. Возможно произошёл технический сбой. Данные лучше уточнить у заказчика.\
    Так как общий трудовой стаж в колонке 'days_employed' не влияет на выполнение поставленных задач, данную колонку заполнил медианным значением.\
    Необходимо заменить отрицательные значения на положительные.\
    5. Колонка 'dob_years' содержит 101 строку с нулевым значением. Возраст не может равняться нулю.\
    6. В колонках 'education' и 'family_status' содержатся неявные дубликаты, имеющие написание с одинаковым значением, но в разных регистрах.\
    Возможен человеческий фактор при внесении данных в базу с использованием клавиш 'Shift' и 'Caps Lock'.\
    Необходимо привести все значения к единому регистру.\
    7. Колонка 'gender' содержит одну строку со странным пропуском 'XNA'.\
    Так как значение единичное, то можем не принимать во внимание.\
    8. Колонка 'debt' принимает два значения: 0 и 1.\
    9. Колонка 'purpose' содержит неявные дубликаты с одинаковым смыслом, которые необходимо удалить.\
    Причиной появления дубликатов возможно является отсутствие конкретного перечня для выбора целевых займов при заполнении анкеты.</b></span></div>

## Изменение аномалий и типов данных

In [88]:
# Изменение отрицательных значений в 'days_employed' на
# положительные в соответствии с источником дохода в 'income_type'
# Создание функции для замены отрицательных значений положительными:
# 1. Объявление функции drop_minus()
def drop_minus(columns):
    # 2. Вложенный цикл для перебора названий
    # колонок из списка drop_minus_columns
    for value in columns:
        # 2.1. Замена отрицательных значений в колонках положительными
        df.loc[df[value] < 0, value] = df[value] * -1
    display(df.head())

In [89]:
# Список колонок с отрицательными значениями
drop_minus_columns = ['days_employed']

In [90]:
# Применение функции drop_minus() к списку drop_minus_columns для
# получения положительных значений с обработкой ошибок try-except
try:
    drop_minus(drop_minus_columns)
except:
    print('Проверь алгоритм работы функции "drop_minus()"!')

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [91]:
# Замена нулевых значений в 'dob_years' средним значением 
# в соответствии с источником дохода в 'income_type'
# Создание функции для замены нулевых значений средним значением:
# 1. Объявление функции to_mean()
def to_mean(columns):
    # 2. Вложенный цикл для перебора названий
    # колонок из списка dob_years_columns
    for value in columns:
        # 2.1. Создание списка уникальных видов дохода
        income_type_list = df['income_type'].unique()
        # 3. Вложенный цикл второго периода для перебора
        # видов дохода из списка income_type_list
        # и формирования датафреймов при соблюдении
        # условий логической индексации с дальнейшим
        # вычислением средних значений в необходимых
        # столбцах value полученных датафреймов
        for string in income_type_list:
            # 3.1. Формирование датафреймов с помощью
            # логической индексации по значению string
            df_string = df[df['income_type'] == string]
            # 3.2. Получение среднего значения столбца
            # value в сформированном датафрейме
            df_mean = df_string[value].mean()
            # 3.3. Замена нулевых значений в колонках
            # value полученными средними значениями
            df.loc[(df[value] == 0) & (df['income_type'] == string), value] = df_mean
    df[value] = df[value].astype('int64')
    display(df[value].describe())

In [92]:
# Список колонок с нулевыми значениями возраста
dob_years_columns = ['dob_years']

In [93]:
# Применение функции to_mean() к списку dob_years_columns для
# замены нулевых значений с обработкой ошибок try-except
try:
    to_mean(dob_years_columns)
except:
    print('Проверь алгоритм работы функции "to_mean()"!')

count    21525.000000
mean        43.495238
std         12.230322
min         19.000000
25%         34.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [94]:
# Изменение типа данных с вщественного float64 на числовой int64 в
# колонках 'days_employed' и 'total_income' внешним циклом для
# перебора названий колонок из списка columns_list
for value in columns_list:
    df[value] = df[value].astype('int64')

In [95]:
# Вывод информации
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


## Удаление дубликатов

In [96]:
# Замена отрицательного количества детей в колонке 'children'
# с помощью логической индексации
df.loc[df['children'] == -1, 'children'] = 1

In [97]:
# Вывод списка неявных дубликатов для проверки
df['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

In [98]:
# Удаление неявных дубликатов в 'education' и 'family_status'
# понижением регистра с использованием внешнего цикла и
# выводом нового списка неявных дубликатов для проверки
drop_unique_duplicates = ['education', 'family_status']
for column in drop_unique_duplicates:
    df[column] = df[column].str.lower()
    display(df[column].value_counts())

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [99]:
# Удаление явных дубликатов с перезаписью индексации и выводом на экран
df = df.drop_duplicates().reset_index(drop=True)

In [100]:
# Вывод датафрейма после удаления явных дубликатов
df

children  days_employed  dob_years education  education_id  \
0             1           8437         42    высшее             0   
1             1           4024         36   среднее             1   
2             0           5623         33   среднее             1   
3             3           4124         32   среднее             1   
4             0         340266         53   среднее             1   
...         ...            ...        ...       ...           ...   
21449         1           4529         43   среднее             1   
21450         0         343937         67   среднее             1   
21451         1           2113         38   среднее             1   
21452         3           3112         38   среднее             1   
21453         2           1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21449  гражданский брак                 1      F   компаньон     0   
21450   женат / замужем                 0      F   пенсионер     0   
21451  гражданский брак                 1      M   сотрудник     1   
21452   женат / замужем                 0      M   сотрудник     1   
21453   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21449        224791             операции с жильем  
21450        155999          сделка с автомобилем  
21451         89672                  недвижимость  
21452        244093  на покупку своего автомобиля  
21453         82047         на покупку автомобиля  

[21454 rows x 12 columns]

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 3:</b></span>

<span style="color:steelblue"><font size="2"><b>После удаления неявных дубликатов в колонках 'education' и 'family_status' были удалены явные дубликаты, количество которых увеличилось до 71.</b></span></div>

## Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма

In [101]:
# Создание нового датафрейма df_education из уникальных
# значений с удалением дубликатов и сбросом индексации
df_education = df.loc[:, ['education', 'education_id']].drop_duplicates().reset_index(drop=True)

In [102]:
# Вывод созданного датафрейма на экран
df_education

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [103]:
# Создание нового датафрейма df_family_status из уникальных
# значений с удалением дубликатов и сбросом индексации
df_family_status = df.loc[:, ['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)

In [104]:
# Вывод созданного датафрейма на экран
df_family_status

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

In [105]:
# Декомпозиция исходного датафрейма удалением
# колонок 'education' и 'family_status'
df = df.loc[:, 
     ['children',
      'days_employed',
      'dob_years',
      'education_id',
      'family_status_id',
      'gender',
      'income_type',
      'debt',
      'total_income',
      'purpose']
]

In [106]:
# Вывод на экран нового датафрейма df
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1           8437         42             0                 0   
1             1           4024         36             1                 0   
2             0           5623         33             1                 0   
3             3           4124         32             1                 0   
4             0         340266         53             1                 1   
...         ...            ...        ...           ...               ...   
21449         1           4529         43             1                 1   
21450         0         343937         67             1                 0   
21451         1           2113         38             1                 1   
21452         3           3112         38             1                 0   
21453         2           1984         40             1                 0   

      gender income_type  debt  total_income                       purpose  
0          F   сотрудник     0        253875                 покупка жилья  
1          F   сотрудник     0        112080       приобретение автомобиля  
2          M   сотрудник     0        145885                 покупка жилья  
3          M   сотрудник     0        267628    дополнительное образование  
4          F   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21449      F   компаньон     0        224791             операции с жильем  
21450      F   пенсионер     0        155999          сделка с автомобилем  
21451      M   сотрудник     1         89672                  недвижимость  
21452      M   сотрудник     1        244093  на покупку своего автомобиля  
21453      F   сотрудник     0         82047         на покупку автомобиля  

[21454 rows x 10 columns]

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 4:</b></span>

<span style="color:steelblue"><font size="2"><b>1. Сформированы два датафрейма с уникальными значениями.\
    1.1. Датафрейм с информацией об образовании 'df_education' содержит 5 категорий.\
    1.2. Датафрейм с информацией о семейном статусе 'df_family_status' так же содержит 5 категорий.\
    2. Декомпозиция исходного датафрейма выполнена. Колонки 'education' и 'family_status' успешно удалены.</b></span></div>

## Категоризация дохода

In [107]:
# Создание колонки 'total_income_category' с категориями дохода
# при помощи функции total_income_category()
# 1. Объявление функции total_income_category()
def total_income_category(total_income):
    # 2. Логические условия для присвоения буквенной категории
    # в зависимости от уровня дохода
    if total_income in range(0, 30000):
        return 'E'
    if total_income in range(30001, 50000):
        return 'D'
    if total_income in range(50001, 200000):
        return 'C'
    if total_income in range(200001, 1000000):
        return 'B'
    return 'A'

In [108]:
# Объединение данных в буквенные категории на основании
# колонки 'total_income' с проверкой ошибок try-except
# и выводом на экран
try:
    df['total_income_category'] = df['total_income'].apply(total_income_category)
    display(df)
except:
    print('Проверь алгоритм работы функции "total_income_category()"!')

children  days_employed  dob_years  education_id  family_status_id  \
0             1           8437         42             0                 0   
1             1           4024         36             1                 0   
2             0           5623         33             1                 0   
3             3           4124         32             1                 0   
4             0         340266         53             1                 1   
...         ...            ...        ...           ...               ...   
21449         1           4529         43             1                 1   
21450         0         343937         67             1                 0   
21451         1           2113         38             1                 1   
21452         3           3112         38             1                 0   
21453         2           1984         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21449      F   компаньон     0        224791             операции с жильем   
21450      F   пенсионер     0        155999          сделка с автомобилем   
21451      M   сотрудник     1         89672                  недвижимость   
21452      M   сотрудник     1        244093  на покупку своего автомобиля   
21453      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category  
0                         B  
1                         C  
2                         C  
3                         B  
4                         C  
...                     ...  
21449                     B  
21450                     C  
21451                     C  
21452                     B  
21453                     C  

[21454 rows x 11 columns]

In [109]:
# Вывод на экран информации о категориях и
# количестве заёмщиков в данных категориях
df['total_income_category'].value_counts()

C    16015
B     5042
D      350
A       25
E       22
Name: total_income_category, dtype: int64

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 5:</b></span>

<span style="color:steelblue"><font size="2"><b>1. Функция 'total_income_category()' работает.\
    2.1. В датафрейме создана новая колонка 'total_income_category'.\
    2.2. Согласно полученным данным 74% заёмщиков относятся к категории 'C' с доходом от 50001 до 200000.</b></span></div>

## Категоризация целей кредита

In [110]:
# Создание колонки 'purpose_category' с категориями
# целей займов при помощи функции purpose_category()
# 1. Объявление функции purpose_category()
def purpose_category(purpose):
    if 'автомоб' in purpose:
        return 'операции с автомобилем'
    if ('недвиж' in purpose or
        'жиль' in purpose):
        return 'операции с недвижимостью'
    if 'свадь' in purpose:
        return 'проведение свадьбы'
    if 'образов' in purpose:
        return 'получение образования'
    return 'проверь ключевые слова и правильность логических условий'

In [111]:
# Объединение данных в категории на основании
# колонки 'purpose' с проверкой ошибок try-except
# и выводом на экран
try:
    df['purpose_category'] = df['purpose'].apply(purpose_category)
    display(df)
except:
    print('Проверь алгоритм работы функции "purpose_category()"!')

children  days_employed  dob_years  education_id  family_status_id  \
0             1           8437         42             0                 0   
1             1           4024         36             1                 0   
2             0           5623         33             1                 0   
3             3           4124         32             1                 0   
4             0         340266         53             1                 1   
...         ...            ...        ...           ...               ...   
21449         1           4529         43             1                 1   
21450         0         343937         67             1                 0   
21451         1           2113         38             1                 1   
21452         3           3112         38             1                 0   
21453         2           1984         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21449      F   компаньон     0        224791             операции с жильем   
21450      F   пенсионер     0        155999          сделка с автомобилем   
21451      M   сотрудник     1         89672                  недвижимость   
21452      M   сотрудник     1        244093  на покупку своего автомобиля   
21453      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category          purpose_category  
0                         B  операции с недвижимостью  
1                         C    операции с автомобилем  
2                         C  операции с недвижимостью  
3                         B     получение образования  
4                         C        проведение свадьбы  
...                     ...                       ...  
21449                     B  операции с недвижимостью  
21450                     C    операции с автомобилем  
21451                     C  операции с недвижимостью  
21452                     B    операции с автомобилем  
21453                     C    операции с автомобилем  

[21454 rows x 12 columns]

In [112]:
# Вывод на экран информации о категориях и
# количестве заёмщиков в данных категориях
df['purpose_category'].value_counts()

операции с недвижимостью    10811
операции с автомобилем       4306
получение образования        4013
проведение свадьбы           2324
Name: purpose_category, dtype: int64

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 6:</b></span>

<span style="color:steelblue"><font size="2"><b>1. Функция 'purpose_category()' работает.\
    2.1. В датафрейме создана новая колонка 'purposee_category'.\
    2.2. Согласно полученным данным 50% займов относятся к категории 'операции с недвижимостью'.</b></span></div>

## Формирование сводных таблиц

### Зависимость между количеством детей и возвратом кредита в срок

In [113]:
# Формирование сводной таблицы методом pivot_table()
# с индексацией по колонке 'children'
df_pivot_children = df.pivot_table(index=['children'], columns='debt', values='income_type', aggfunc='count')

In [114]:
# Добавление колонки с процентным соотношением по задолженности
df_pivot_children['percent'] = df_pivot_children[1] / (df_pivot_children[0] + df_pivot_children[1]) * 100

In [115]:
# Вывод сводной таблицы с сортировокй по процентам
df_pivot_children.sort_values(by='percent')

debt            0       1    percent
children                            
0         13028.0  1063.0   7.543822
3           303.0    27.0   8.181818
1          4410.0   445.0   9.165808
2          1858.0   194.0   9.454191
4            37.0     4.0   9.756098
20           68.0     8.0  10.526316
5             9.0     NaN        NaN

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 7:</b></span>

<span style="color:steelblue"><font size="2"><b>1. В сводной таблице по колонке 'children' с количеством детей присутствует значение NaN(Not-a-Number). В связи с этим все значения приобрели вещественный тип данных float64.\
    Появление NaN возможно обусловлено отсутствием задолженности (1) в колонке 'debt' у семей с 5 детьми.\
    2. Количество детей у семьи влияет на общее количество займов: чем меньше детей, тем больше займов.\
    Исключение составляют семьи с аномальным количеством детей 20.\
    3. Нименьшее количество задолженностей по возврату кредитов имеют семьи с количеством детей 0: примерно 7,5% задолженностей от взятых займов этой категорией семей.\
    С увеличением числа детей увеличивается процент задолженностей, за исключением семей с 3 детьми.</b></span></div>

### Зависимость между семейным положением и возвратом кредита в срок

In [116]:
# Формирование сводной таблицы методом pivot_table()
# с индексацией по колонке 'family_status_id'
df_pivot_family = df.pivot_table(index=['family_status_id'], columns='debt', values='income_type', aggfunc='count')

In [117]:
# Добавление колонки с процентным соотношением по задолженности
df_pivot_family['percent'] = df_pivot_family[1] / (df_pivot_family[0] + df_pivot_family[1]) * 100

In [118]:
# Вывод сводной таблицы с сортировокй по процентам
df_pivot_family.sort_values(by='percent')

debt                  0    1   percent
family_status_id                      
2                   896   63  6.569343
3                  1110   85  7.112971
0                 11408  931  7.545182
1                  3763  388  9.347145
4                  2536  274  9.750890

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 8:</b></span>

<span style="color:steelblue"><font size="2"><b>1. Семейный статус влияет на количество задолженностей.\
    2. Наименьшее количество задолженностей у категории 2 и 3 - вдовцов и разведённых. Возможно люди, состоявшие в браке, меньше берут займы, так как рассчитывают только на свои доходы.\
    3. Наибольшее количество займов со средним показателем задолженности у семей, состоящих в браке - категория 0. Возможно это связано с общим бюджетом семьи и взаимной поддержкой.\
    4. Наибольший процент задолженности наблюдается у семей, нескреплённых брачными узами - категория 1, и одиночек - категория 4. Возможно у них слабая взаимная поддержка и также переоценённая платёжеспособность.</b></span></div>

### Зависимость между уровнем дохода и возвратом кредита в срок

In [119]:
# Формирование сводной таблицы методом pivot_table()
# с индексацией по колонке 'total_income_category'
df_pivot_income = df.pivot_table(index=['total_income_category'], columns='debt', values='income_type', aggfunc='count')

In [120]:
# Добавление колонки с процентным соотношением по задолженности
df_pivot_income['percent'] = df_pivot_income[1] / (df_pivot_income[0] + df_pivot_income[1]) * 100

In [121]:
# Вывод сводной таблицы с сортировокй по процентам
df_pivot_income.sort_values(by='percent')

debt                       0     1   percent
total_income_category                       
D                        329    21  6.000000
B                       4686   356  7.060690
A                         23     2  8.000000
C                      14655  1360  8.492039
E                         20     2  9.090909

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 9:</b></span>

<span style="color:steelblue"><font size="2"><b>1. Ситуация с влиянием уровня дохода на количество задолженностей неоднозначная.\
    2. Наименьшее количество задолженностей в категории с предпоследним уровнем дохода - категории D, но и займов в данной категории небольшое количество.\
    3. У категорий с большими доходами - категрии A и B, задолженности так же имеются, хоть и с меньшим процентом. Большие доходы предполагают большие расходы и связанные с ними риски.\
    4. В категории с наибольшим количеством займов - категории C, наблюдается предпоследний по величине процент по задолженностям.\
    5. В категории с наименьшим доходом - категории E, самый высокий процент задолженности.</b></span></div>

### Влияние целей кредита на его возврат в срок

In [122]:
# Формирование сводной таблицы методом pivot_table()
# с индексацией по колонке 'purpose_category'
df_pivot_purpose = df.pivot_table(index=['purpose_category'], columns='debt', values='income_type', aggfunc='count')

In [123]:
# Добавление колонки с процентным соотношением по задолженности
df_pivot_purpose['percent'] = df_pivot_purpose[1] / (df_pivot_purpose[0] + df_pivot_purpose[1]) * 100

In [124]:
# Вывод сводной таблицы с сортировокй по процентам
df_pivot_purpose.sort_values(by='percent')

debt                          0    1   percent
purpose_category                              
операции с недвижимостью  10029  782  7.233373
проведение свадьбы         2138  186  8.003442
получение образования      3643  370  9.220035
операции с автомобилем     3903  403  9.359034

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>Промежуточные выводы 10:</b></span>

<span style="color:steelblue"><font size="2"><b>1. Операции с недвижимостью занимают большую часть займов. В виду высокой стоимости недвижимости заёмщики подходят основательно к данной операции, поэтому и процент задолженностей самый низкий.\
    2. Проведение свадеб не несёт огромных расходов. У молодожёнов свадьбы часто выстраиваются на окупаемость. Возможно поэтому процент задолженностей занимает второе место.\
    3. В операциях с получением образования возможно займы берут люди с низким доходом, сами студенты с нестабильным источником заработка.\
    4. Самый высокий процент задолженностей в опреациях с автомобилями. Возможно такой займ чаще берут спонтанно, находясь в автосалоне. Также автокредиты берут молодые люди без опыта вождения. Не исключаю влияние высокого уровня аварийности на дорогах.</b></span></div>

## Общий вывод

<div class="alert alert-block alert-warning">

<span style="color:green"><font size="3"><b>В ходе исследования выявлена зависимость количества детей, семейного положения, дохода, цели займа с количеством задолженностей по платежу.\
    В целом больших интервалов по проценту задолженностей между категориями не выявлено. Риски существуют во всех категориях.\
    Идеальным заёмщиком является вдовец/вдова без детей, с уровнем дохода 30001-50000 либо 200001-1000000, нацеленный на операцию с недвижимостью, и которого в трудную минуту может поддержать другой вдовец/вдова.**</b></span></div>